### ISSUE TITLE
CompartmentMapper - cjs exports defined through defineProperty are not handled correctly

### ISSUE THREAD

**Issue by kumavis** [endojs/endo#885]
example defineProperty based exports in the wild https://npmfs.com/package/ethereumjs-util/7.1.0/dist/index.js#L4

in this case this problem might be worked around by convincing the package author to also distribute esm format

**Comment by kumavis** [comment]
issue on ethereumjs https://github.com/ethereumjs/ethereumjs-monorepo/issues/1468

**Comment by naugtur** [comment]
configurable would have to be true: https://github.com/endojs/endo/blob/02731651bb9ad9d895e1aa740e2634600ee74404/packages/ses/src/module-instance.js#L68

**Comment by naugtur** [comment]
I've got a solution to this using a proxy in parse-cjs.js:

```js

    const exportsDoubleProxy = new Proxy(
      {},
      {
        get(target, prop) {
          return moduleExports[prop] || target[prop]; // this makes things like exports.hasOwnProperty() work. Does it break hardening though?
        },
        set(target, prop, value) {
          moduleExports[prop] = value;
          return true;
        },
      },
    );

    let exportsReferenceCopy = exportsDoubleProxy;

    const module = freeze({
      get exports() {
        return exportsReferenceCopy;
      },
      set exports(value) {
        exportsReferenceCopy = value;
      },
    });

( ...)

    functor(
      require,
      exportsReferenceCopy,
      module,
      location, // __filename
      new URL('./', location).toString(), // __dirname
    );
    if (exportsReferenceCopy !== exportsDoubleProxy) {
      moduleExports.default = exportsReferenceCopy;
    }
```

note that the proxy reference is never given to anyone other than the module itself. Everything written to it is being passed on to moduleExports as it happens, so even cjs modules with a cyclic dependency will synchronously have the right references available.

@kriskowal  is this a bad idea? Also, why did compartment-shim avoid using a proxy?

**Comment by kriskowal** [comment]
This seems like a good idea, or at least the beginning of one. The Compartment shim will eventually need to provide a Proxy for the module environment record. I generally avoid using Proxy until it’s clearly necessary for performance reasons.

**Comment by naugtur** [comment]
This got _even more_ fixed by #1074 and that's where the relevant implementation to address this issue is now coming from
